### DISCUSSION TITLE
Right function for Iterable<Either<E, A>>

### DISCUSSION THREAD

**Discussion by serras** [arrow-kt/arrow#2861] [Ideas]
During the discussion in https://github.com/arrow-kt/arrow/pull/2795, the issue of a good function that works over `Iterable<Either<E, A>` was raised by @pakoito. Here are some of the ideas gathered, so we can discuss them separately from that PR.

### 1. `Iterable<Either<E, A>>.sequence(): Either<E, List<A>>`

This function is very similar to the `sequence` found in other languages; it returns the original list of elements, or bails out if any of them is `Left`. Note that this only returns the **first** error.

### 2. `Iterable<Either<E, A>>.flattenOrAccumulate(): Either<NonEmpty<E>, List<A>>`

This was suggested by @nomisRev in an internal conversation. The idea is that it works like `sequence` above, but doesn't stop at the first error. This suits a bit better with the new API in which we try to maintain errors as far as possible, as we used to do with `Validated`.

### 3. `suspend Iterable<Either<E, A>>.bindAll(): List<A>`

This would be just a shortcut to `map { it.bind() }`. The idea is to stay within the DSL, but has the disadvantage that you need to write `either { list.bindAll() }` to recover the behavior of `sequence`; it also stops on the **first** error.

**Comment by serras** [comment]
This PR https://github.com/arrow-kt/arrow/pull/2872 proposes yet another way to introduce an accumulating function, in this case generic over every `Raise`.

**Comment by nomisRev** [comment]
Here is a use-case for the 2nd signature, https://github.com/arrow-kt/arrow/pull/2863/files#diff-74bf2a702683ff9b5377a1d02d00a2708862d90d5d5dc25e72f675283ee14b1cR73. Name is up for debate, but it's needed to derive the _parallel_ version `mapOrAccumulate`.

After some discussion in the last bi-weekly Arrow meeting, I am in favour of adding both options `2` and `3`. IMO `bindAll` makes more sense than having `sequence`.

**Comment by kyay10** [comment]
`bindAll` now exists, and it accumulates if called within `RaiseAccumulate`
